File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
//@Author: KING-SEN
// Programming Language Used: C++

class Solution {
public:
    int trap(vector<int>& height) {
        int start = 0;
        int end = height.size() - 1;
        
        int maxleft = height[start];
        int maxright = height[end];
        
        int answer = 0;
        
        while (start < end) {
            if (maxleft <= maxright) {
                start++;
                maxleft = max(maxleft, height[start]);
                answer += maxleft - height[start];
            } else {
                end--;
                maxright = max(maxright, height[end]);
                answer += maxright - height[end];
            }
        }
            
        return answer;
    }
};

// Time Complexity: O(N), where N is the length of the array
// Space Complexity: O(1), Here we use constant space

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left, right = 0, len(height) - 1
        leftmax, rightmax = height[0], height[-1]
        res = 0
        while left < right:
            if leftmax <= rightmax:
                left += 1
                leftmax = max(leftmax, height[left])
                res += leftmax - height[left]
            else:
                right -= 1
                rightmax = max(rightmax, height[right])
                res += rightmax - height[right]
        return res
            
            